### PR TITLE
cbn: better handling of volatile constants

### LINKED ISSUE #15567: cbn unfolding regressions in 8.15
#### Description of the problem
The following example shows `cbn` unfolding a term which was not unfolded in 8.13 and 8.14. Judging by the documentation of `Arguments`, the term should not be unfolded.
```coq
From Coq Require Import List.
Fixpoint big_opL {A} f (xs : list A) : Prop :=
  match xs with
  | nil => True
  | x :: xs => (f 0 x) /\ (big_opL (fun n => f (S n)) xs)
  end.
Arguments big_opL {_} _ !xs /.
Goal forall A f xs, @big_opL A f xs.
  induction xs; [admit|]; cbn.
(*
  f 0 a /\
  (fix big_opL (A0 : Type) (f0 : nat -> A0 -> Prop) (xs0 : list A0) {struct xs0} : Prop :=
     match xs0 with
     | nil => True
     | x :: xs1 => f0 0 x /\ big_opL A0 (fun n : nat => f0 (S n)) xs1
     end) A (fun n : nat => f (S n)) xs 
*)
```

/cc @rlepigre

#### Coq Version
8.15

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #15567

### CODE REVIEW

File: tactics/cbn.ml [hunk 1]
```diff
@@ -38,34 +39,34 @@ module Cst_stack = struct
 - constant applied to params = term in head applied to args
 - there is at most one arguments with an empty list of args, it must be the first.
 - in args, the int represents the indice of the first arg to consider *)
-  type t = (constr * constr list * (int * constr array) list)  list
+  type t = (constr * constr list * (int * constr array) list * volatile)  list
 
   let empty = []
-  let is_empty = CList.is_empty
+  let all_volatile = CList.for_all (fun (_,_,_,{volatile}) -> volatile)
```
Comment (SkySkimmer) [inline tactics/cbn.ml:45]: CI errors for hott and relation_algebra in the first run seem to have been because of this mising change.

**Comment by JasonGross** [comment]
> @JasonGross that iris minimization looks bugged

@SkySkimmer Indeed, thanks for the tag, I'm now tracking this at https://github.com/coq/coq/pull/14748#issuecomment-1034527240 and https://github.com/coq-community/run-coq-bug-minimizer/issues/8.  It looks like the sanity check is failing for reasons that are not clear to me...

**Comment by SkySkimmer** [comment]
Split off the simpl nomatch part to https://github.com/coq/coq/pull/15657

File: tactics/cbn.ml [hunk 2]
```diff
@@ -38,34 +40,34 @@ module Cst_stack = struct
 - constant applied to params = term in head applied to args
 - there is at most one arguments with an empty list of args, it must be the first.
 - in args, the int represents the indice of the first arg to consider *)
-  type t = (constr * constr list * (int * constr array) list)  list
+  type t = (constr * constr list * (int * constr array) list * volatile)  list
```
Comment (ppedrot) [inline tactics/cbn.ml:43]: I know it requires a larger diff, but would it be possible to get a record type here? This is starting to be really to big for its own good.

Comment (SkySkimmer) [reply tactics/cbn.ml:43]: The pattern matchings get pretty painful
I can try again if you insist

Comment (ppedrot) [reply tactics/cbn.ml:43]: Nah, I've tried that at some point as well and let the patch die. It was indeed annoying. Better merge this as-is then.

**Review by ppedrot — APPROVED** [review]

**Pull request merged**